### ISSUE TITLE
Enable or reimplement meta-data test

### ISSUE THREAD

**Issue by k-rus** [Starcounter/Starcounter#3042]
Part of QueryProcessingTest is test of populated meta-data. It is good to enable it to see that meta-data are populated in expected way. It is important to fix it that it does not fail on included so classes.
It is important that the test checks not just that meta-data was created, but also that it was not created more than should be.

**Comment by k-rus** [comment]
After discussion with @dan31 the issue is assigned to @per-samuelsson

**Comment by per-samuelsson** [comment]
Where can I find this test?

**Comment by k-rus** [comment]
It is in QueryProcessingTest.  The test function called `MetadataTest.TestPopulatedMetadata()`, which is currently disabled in the main program of the test.

**Comment by per-samuelsson** [comment]
> t is in QueryProcessingTest. The test function called MetadataTest.TestPopulatedMetadata(), which is currently disabled in the main program of the test.

Its [this](https://github.com/Starcounter/Starcounter/blob/pnext/src/Samples/QueryProcessingTest/MetadataTest.cs), right?

I just how no idea how I should verify and own this stuff, when I'm not involved in metadata generation <i>**at all**</i>? 

How should I know the type of low-level instricts that your checks do? Like that if I do a query on `DbPrimitiveType`, the first hit should be `string`, or that the query `select c from materializedcolumn c where c.name = "inherits"` should return a COUNT of 4?!

I really don't understand what it is expected I should do here?

**Comment by k-rus** [comment]
Now, it is enough that the meta-data test checks if corresponding instances are created in `Metadata.ClrClass`, `MappedProperty` and `CodePropery` for some user-defined classes used in the test. And that there is no duplicate instances are created. And that the instances are correctly referencing `RawView` and `RawColumn` and no duplicates there.

Now, there is no materialised tables any more. Testing `RawView` and `RawColumn` is not so important any more, since they are already in use and failure will be noticed before this test. No test on total number of records in meta-tables needed, since it breaks on SO.

If you don't feel to do this task, let us know and re-assign from it.

**Comment by per-samuelsson** [comment]
> Now, it is enough that the meta-data test checks if corresponding instances are created in Metadata.ClrClass, MappedProperty and CodePropery for some user-defined classes used in the test. And that there is no duplicate instances are created. And that the instances are correctly referencing RawView and RawColumn and no duplicates there.

Thanks. Thats a whole different story.

Generally, **my opinion** (and I've told you that earlier) is that we should transforrm these tests to work based on what the **end user / consumer** would expect. For me, testing *implementation details* is the task of unit testing (if ever; even better, I think unit testing of API's + code coverage is a better approach). 

Regression tests like these should validate stuff like "If I define a class with two fields, I want to assure there is a single instance of `ClrClass` with that name and those two fields represented as `MappedProperty`". Not that there are 4 columns in total in some underlying table.

Again, thats **my opinion** of how we should design these tests. That gives us the freedom to tweak the implementation, and still assuring the expected behaviour.

> If you don't feel to do this task, let us know and un-assign from it.

If I'm free to switch the focus of the test to what I describe above, I'll gladly do this.

**Comment by k-rus** [comment]
> For me, testing implementation details is the task of unit testing (if ever; even better, I think unit testing of API's + code coverage is a better approach).
> 
> Regression tests like these should validate stuff like "If I define a class with two fields, I want to assure there is a single instance of ClrClass with that name and those two fields represented as MappedProperty". Not that there are 4 columns in total in some underlying table.
> 
> Again, thats my opinion of how we should design these tests. That gives us the freedom to tweak the implementation, and still assuring the expected behaviour.

I agree with your vision. 
Since the test is about the database state, it is not just unit test, but more like functional system test. Such test will need to be put inside bootstrapper, but then it will be assurance not a test, or it should take weaver result as input to match with the meta-data state. So I guess it is not so simple task.



> Generally, my opinion (and I've told you that earlier) is that we should transforrm these tests to work based on what the end user / consumer would expect.

The main consumer is new SQL processor. Starcounter's users are also currently consumers, but later they will be given views, which will be designed for them.

Btw, I forgot to mention that index meta-data are already tested in another test.

**Comment by per-samuelsson** [comment]
> The main consumer is new SQL processor.

Yes, but still, why does this consumer care if there are 4 columns named "Inherited" in the column table? 

My point is those kind of test serve very little purpuse. They make it hard to refactor the **implementation** and yet provide no real need to consumers, b/c consumers just wouldn't know about them. **And** they make it really hard for someone else to write/verify, since you need to be the low-level implemetor of the module to know about such things. That's why it would make it impossible for me to do this task.

But with all being said, my understanding now is that I'm allowed to redesign these tests in the direction I propose, and then I'll gladly do this.

**Comment by k-rus** [comment]
> But with all being said, my understanding now is that I'm allowed to redesign these tests in the direction I propose, and then I'll gladly do this.

Correct. Great!

> Yes, but still, why does this consumer care if there are 4 columns named "Inherited" in the column table?
> 
> My point is those kind of test serve very little purpuse. They make it hard to refactor the implementation and yet provide no real need to consumers, b/c consumers just wouldn't know about them. And they make it really hard for someone else to write/verify, since you need to be the low-level implemetor of the module to know about such things.

I agree. It was bad idea. I used my knowledge of C# classes, which I wrote, and moved this knowledge into the test.

**Comment by k-rus** [comment]
Related bug report in https://github.com/Starcounter/Starcounter/issues/2938
